leopard/wrapper.nim

```diff
@@ -0,0 +1,293 @@
+## originalCount: Number of original data buffers provided.
+## recoveryCount: Number of desired recovery data buffers.
+## workCount:     Number of work data buffers, from leoEncodeWorkCount().
+## originalData:  Array of pointers to original data buffers.
+## workData:      Array of pointers to work data buffers.
+##
+## The sum of originalCount + recoveryCount must not exceed 65536.
+## The recoveryCount <= originalCount.
+##
+## The value of bufferBytes must be a multiple of 64.
+## Each buffer should have the same number of bytes.
+## Even the last piece must be rounded up to the block size.
+##
+## Returns LeopardSuccess on success.
+## The first set of recoveryCount buffers in workData will be the result.
+## Returns other values on errors.
+
+proc leoEncode*(
+  bufferBytes: uint64,   ## Number of bytes in each data buffer
+  originalCount: cuint,  ## Number of originalData[] buffer pointers
+  recoveryCount: cuint,  ## Number of recovery data buffer pointers
+                         ## (readable post-call from start of workData[])
+  workCount: cuint,      ## Number of workData[] buffer pointers
+  originalData: pointer, ## Array of pointers to original data buffers
+  workData: pointer,     ## Array of pointers to work data buffers
+): LeopardResult {.leo, importc: "leo_encode".}
+
+
+## -----------------------------------------------------------------------------
+## Decoder API
+
+## leoDecodeWorkCount()
+##
+## Calculate the number of work data buffers to provide to leoDecode().
+##
+## The sum of originalCount + recoveryCount must not exceed 65536.
+##
+## Returns the workCount value to pass into leoDecode().
+## Returns 0 on invalid input.
+
+func leoDecodeWorkCount*(originalCount, recoveryCount: cuint): cuint
+  {.leo, importc: "leo_decode_work_count".}
+
+## leoDecode()
+##
+## Decode original data from recovery data.
+##
+## bufferBytes:   Number of bytes in each data buffer.
+## originalCount: Number of original data buffers provided.
+## recoveryCount: Number of recovery data buffers provided.
+## workCount:     Number of work data buffers, from leoDecodeWorkCount().
+## originalData:  Array of pointers to original data buffers.
+## recoveryData:  Array of pointers to recovery data buffers.
+## workData:      Array of pointers to work data buffers.
+##
+## Lost original/recovery data should be set to NULL.
+##
+## The sum of recoveryCount + the number of non-NULL original data must be at
+## least originalCount in order to perform recovery.
+##
+## Returns LeopardSuccess on success.
+## Returns other values on errors.
+
+proc leoDecode*(
+  bufferBytes: uint64,   ## Number of bytes in each data buffer
+  originalCount: cuint,  ## Number of originalData[] buffer pointers
+  recoveryCount: cuint,  ## Number of recoveryData[] buffer pointers
+  workCount: cuint,      ## Number of workData[] buffer pointers
+  originalData: pointer, ## Array of pointers to original data buffers
```

Comment:
This has to be `ptr pointer` or even `ptr UncheckedArray[byte]`

Reply:
I've changed the `pointer` parameters in `leo/wrapper.nim` to `ptr pointer`.

Tests pass with just `pointer`, but they also pass with `ptr pointer` As discussed during last Thursday's call, the latter may be more correct re: the code being wrapped and just `pointer` works correctly by coincidence.

Reply:
Yeah, it's by coincidence because you have a contiguous chunk of address space. This is a good reference on pointers, which are relevant in all contexts, even higher level software - https://en.wikipedia.org/wiki/Pointer_(computer_programming)